src/guidellm2pcp/guidellm2pcp.1
'\"macro stdmacro
.\"
.\" Copyright (C) 2025 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH GUIDELLM2PCP 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3guidellm2pcp\f1 \- GuideLLM benchmark results importer
.SH SYNOPSIS
\fBguidellm2pcp\fP
[\fB\-hv\fP]
[\fB\-a\fP \fIarchive\fP]
[\fB\-H\fP \fIhostname\fP]
\fIresults\fP
[...]
.SH DESCRIPTION
.B guidellm2pcp
is a performance metrics importer tool that creates a PCP archive
from the JSON results of a GuideLLM benchmark.
.PP
An archive spanning the time for which the benchmark and its run
iterations were active.
Discrete PCP metrics are created for each recorded aspect of a run,
and logged once at the start and once at the end of each iteration,
followed by a mark record.
.PP
This allows PCP tools to report on these metrics in a meaningful way
alongside other system (kernel, hardware, vLLM, and so on) metrics
gathered during the benchmark.
.PP
The only mandatory command line argument is the JSON
.I results
file from a GuideLLM invocation.
This JSON can optionally be wrapped with additional Model Furnace
metadata, and if this is detected additional PCP labels will be
extracted and stored (such as model and inference server names).
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-a\fR \fIarchive\fR, \fB\-\-archive\fR=\fIarchive\fR
Specified the name of the output PCP
.I archive
to be produced.
By default, an archive name using the timestamp of the start of
the benchmark run will be used.
.TP
\fB\-H\fR \fIhostname\fR, \fB\-\-hostname\fR=\fIhostname\fR
Sets the
.I hostname
that will be stamped into the PCP archive label header.
.TP
\fB\-h\fR, \fB\-\-help\fR
Display usage message and exit.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Report progress as each run is identified and its values saved.
.SH SEE ALSO
.BR PCPIntro (1),
.BR pminfo (1),
.BR pmrep (1),
.BR LOGARCHIVE (5),
and
.BR PMNS (5).

.\" control lines for scripts/man-spell
.\" +ok+ GuideLLM EST


CHANGELOG
pcp-7.0.2 (15 October 2025)
    - Currently work-in-progress.

pcp-7.0.1 (21 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: new per-process metrics agent for macOS
	pmdagfs2: Add support for glockfd debugfs metric collection
	pmdalinux: add NFSv4 connected clients and open files metrics
	pmdalinux: additional /proc/net/netstat MPTcp metrics
	pmdabpftrace: Adjust event loop code to work with Python 3.14
    - Client tools and utilities:
	guidellm2pcp: convert GuideLLM benchmark results to PCP archives
	pcp-mpstat: use vuser/vnice metrics and unify CPU utilization values
    - Server-side utilities and log management scripts:
	pmproxy, libpcp_web: ensure correct path used for conflict resolution
	pmlogger, pmie scripts: improvements for BSD date(1) differences
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix accounting within __pmHashCtl after walk-based-delete
	libpcp: fixes for >2GB archive files on 32-bit platforms
	python api: fix datetime-based timestamps in import API
	python api: improve exception handler in import API
    - Build, infrastructure, containers and packaging updates:
	build: automate macOS release and build process on github
	build: add aarch64 archive-analysis container builds on quay+ghcr
	build: debian conditional Build-Depends: for systemd pkgconfig file
	build: packaging list updates for RHEL 10 builds of PCP
	build: add pmproxy to the rpm set needing LOCAL_ONLY port setup
	build: ensure pcp-zeroconf rpm install script succeeds
	build: add zstd package as a hard rpm package dependency
	containers: archive-analysis container exports optional REST API
	containers: switch archive-analysis to using pcp-import-pmseries
	containers: add PCP_DOMAIN_AGENTS to pcp for auto-enabling PMDAs
	containers: make most/all PMDAs available in the pcp container
	containers: update the pmproxy template config in pcp container
    - Documentation:
	docs: fix openapi.yaml issues (pcp.readthedocs.io/en/latest/api)
	docs: add a containers note about alternate registry at ghcr.io
	docs: add a CLAUDE.md file for devs using claude code with PCP

pcp-7.0.0 (1 September 2025)
    - PMDA additions, enhancements and bug fixes:
	pmdaopentelemetry: new exporter for OpenTelemetry metrics
	pmdarocestat: new RDMA over Converged Ethernet (RoCE) metrics
	pmdahdb: new agent exporting SAP HANA metrics
	pmdalinux: additional TCP networking metrics
	pmdalinux: additional kernel memory and hugetlb metrics
	pmdalinux: support Amazon Linux within kernel.uname.distro
	pmdaopenmetrics: improvements to dynamic metric name handling
	pmdaopenmetrics: allow configuration files with epoch times
	pmdapodman: fix container name/command handling in new podman
	pmdahacluster: additional pacemaker node metrics
	pmdasmart: add smart.wwid.* metrics with drive UUID for indom
	pmdammv: support empty instance domains and dynamic map resets
	pmdammv: export instance domain help text when requested
    - Server-side utilities and log management scripts:
	pmlogger, pmproxy: support for a remote archive push webhook;
	  providing central logging without server-side configuration
	pmlogconf: updates to ensure complete sysstat metric coverage
	pmproxy_daily: new daily script for pmlogger-pushed archives
	pmproxy: support both OpenMetrics and OpenTelemetry /metrics
	pmproxy: configurable HTTP headers (e.g. X-Grafana-Device-ID)
	pmlogger_daily: default compression is a dynamic zst/xz choice
	rc_pcp: legacy init.d pmcd+pmlogger wrapper script removed
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp.so.4: default PMAPI supports nanosecond precision times
	libpcp.so.3: provides backwards compatible libpcp interfaces
	pmapi.h: backwards compatible APIs via -DPMAPI_VERSION=2|4
	libpcp: flag for pmproxy to open archives at the last volume
	libpcp: fix lurking __pmOptFetchAdd bug affecting pmlogger
	libpcp_import: protect against a second call to pmiEnd
	python api: fix (de)reference of a tuple in refresh callback
	python api: removed python version 2 support, 3.6 is minimum
    - Client tools and utilities:
	pcp-rocestat: client to report NVIDIA RoCE metrics
	pcp-dstat: fix the --time-adv option with archives
	pcp-htop: sync-up with latest htop.dev code fixes
	pcp-htop: improve pmcd disconnect/restart handling
	pmrep: fix iostat merge percentage formulas
	pmrep: allow for not preserving discrete metric values
	pmrep: add sar-B and sar-r columns, sync with recent sysstat
	pmjson: produce colorful output to the console by default
	pmlogpush: new tool for pushing existing archives to pmproxy
	pcp2opentelemetry: new tool exporting to OpenTelemetry format
	pmseries_import: new tool for archive analysis container
	archive analysis container: removed rebuild of PCP sources
	archive analysis container: refactored to use real packages
	archive analysis container: set Grafana time window from archive
    - Build, infrastructure and packaging updates:
	build: MacOS builds and packaging, welcome back!
	build: removed Qt4 support from code base, 5.6 is minimum
	build: various improvements for reproducible builds
	build: archive-analysis container reworked to use packages
	build: fixes to better support parallel building
	build: fix multilib installs for pcp-libs-devel packages
	build: provide libvirt PMDA on any el7 if py3 modules exist
	build: clean up debian missing Build-Depends: packages
	build: allow -DNDEBUG to be used consistently across distro builds
	build: allow containers of any architecture to be built
	selinux: update policy for AMD and NVIDIA gpu pmdas
	selinux: additional policy needed for pcp_pmie_t using ps
	selinux: add permissions allowing proc_psi_t access
	tmpfiles.d: additions to support better ostree integration
	pcp-zeroconf: use systemd presets to drive service start
    - Documentation:
	man pages: improved function prototypes in SYNOPSIS section
	man pages: updates to PCPCompat(1) describing major changes


src/guidellm2pcp/guidellm2pcp.py
#!/usr/bin/pmpython
#
# Copyright (C) 2025 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# Create a PCP archive from GuideLLM benchmark results JSON file.
#

import argparse
import json
import sys

from datetime import datetime, timezone
from pcp import pmapi, pmi
from cpmapi import (PM_COUNT_ONE, PM_LABEL_CONTEXT,
                    PM_TIME_SEC, PM_TIME_MSEC,
                    PM_TYPE_DOUBLE, PM_TYPE_U64, PM_SEM_DISCRETE,
                    PM_TEXT_PMID, PM_TEXT_ONELINE, PM_TEXT_HELP)

def register_metrics(pcp) -> pmapi.pmInDom:
    domain = 510 # PCP domain number (guidellm import)
    indom = pcp.pmiInDom(domain, 1) # benchmark iterations
    cluster = 0 # incremented for each metric group added
    item = 0 # incremented for each metric within a group

    nounit = pcp.pmiUnits(0, 0, 0, 0, 0, 0)
    unitone = pcp.pmiUnits(0, 0, 1, 0, 0, PM_COUNT_ONE)
    secunits = pcp.pmiUnits(0, 1, 0, 0, PM_TIME_SEC, 0)
    msecunits = pcp.pmiUnits(0, 1, 0, 0, PM_TIME_MSEC, 0)
    psecunits = pcp.pmiUnits(0, 1, 1, 0, PM_TIME_SEC, PM_COUNT_ONE)

    def add_metric(name, datatype, units, oneline=None, text=None) -> None:
        nonlocal item
        pmid = pcp.pmiID(domain, cluster, item)
        pcp.pmiAddMetric(name, pmid, datatype,
                     indom, PM_SEM_DISCRETE, units)
        if oneline:
            pcp.pmiPutText(PM_TEXT_PMID, PM_TEXT_ONELINE, pmid, oneline)
        if text:
            pcp.pmiPutText(PM_TEXT_PMID, PM_TEXT_HELP, pmid, text)
        item += 1

    pmns = 'guidellm.'
    add_metric(pmns + 'duration', PM_TYPE_DOUBLE, secunits,
               'Elapsed time in seconds for each GuideLLM run completed')
    pmns = 'guidellm.run_stats.requests_made.'
    add_metric(pmns + 'successful', PM_TYPE_U64, unitone,
               'Count of requests processed (requests-per-second as a rate)')
    add_metric(pmns + 'errored', PM_TYPE_U64, unitone,
               'Count of failed requests, e.g. invalid input or server issues')
    add_metric(pmns + 'incomplete', PM_TYPE_U64, unitone,
               'Count of incomplete requests, e.g. timeouts or interruptions')
    add_metric(pmns + 'total', PM_TYPE_U64, unitone,
               'Count of all requests started irrespective of how they finish')

    pmns = 'guidellm.run_stats.'
    add_metric(pmns + 'queued_time_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'scheduled_time_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'scheduled_time_sleep_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'worker_start_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'worker_time_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'worker_start_time_targeted_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'request_start_time_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'request_start_time_targeted_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'request_time_delay_avg', PM_TYPE_DOUBLE, secunits)
    add_metric(pmns + 'request_time_avg', PM_TYPE_DOUBLE, secunits)

    pmns = 'guidellm.requests_per_second.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, psecunits,
        'Average number of requests processed per second',
        'Indicates throughput and the systems ability to handle concurrent workloads.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, psecunits,
        'Median number of requests processed per second',
        'Indicates throughput and the systems ability to handle concurrent workloads.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, psecunits,
        'Most frequently occurring number of requests processed per second',
        'Indicates throughput and the systems ability to handle concurrent workloads.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Total number of requests per second data points')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, psecunits,
        'Minimum requests per second')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, psecunits,
        'Maximum requests per second')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of requests per second')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within requests per second')

    pmns = 'guidellm.request_latency.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, msecunits,
        'Average time taken to process requests',
        'A critical metric for evaluating the responsiveness of the system.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, msecunits,
        'Median time taken to process requests',
        'A critical metric for evaluating the responsiveness of the system.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, msecunits,
        'Most frequently occurring time taken to process requests',
        'A critical metric for evaluating the responsiveness of the system.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Total number of time taken to process request data points')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, msecunits,
        'Minimum time taken to process a single request')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, msecunits,
        'Maximum time taken to process a single request')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of time taken to process requests')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within time taken to process requests')

    pmns = 'guidellm.time_to_first_token_ms.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, msecunits,
        'Average time taken to generate the first token of the output',
        'Indicates initial response time of the model, crucial for user-\n' +
        'facing applications.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, msecunits,
        'Median time taken to generate the first token of the output',
        'Indicates initial response time of the model, crucial for user-\n' +
        'facing applications.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, msecunits,
        'Most frequently occurring time taken to generate first output token',
        'Indicates initial response time of the model, crucial for user-\n' +
        'facing applications.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Total number of time taken to generate first output token measures')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, msecunits,
        'Minimum time taken to generate first output token')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, msecunits,
        'Maximum time taken to generate first output token')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of time taken to generate first output token')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within time taken to generate first output token')

    pmns = 'guidellm.time_per_output_token_ms.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, msecunits,
        'Average time taken to generate each output token',
        'Provides a detailed view of the models token generation efficiency.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, msecunits,
        'Median time taken to generate each output token',
        'Provides a detailed view of the models token generation efficiency.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, msecunits,
        'Most frequently occurring time taken to generate each output token',
        'Provides a detailed view of the models token generation efficiency.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Count of time taken to generate each output token measures')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, msecunits,
        'Minimum time taken to generate each output token')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, msecunits,
        'Maximum time taken to generate each output token')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of time taken to generate each output token')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within time taken to generate each output token')

    pmns = 'guidellm.inter_token_latency_ms.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, msecunits,
        'Average time between generating consecutive tokens in the output',
        'Helps assess the smoothness and speed of token generation.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, msecunits,
        'Median time between generating consecutive tokens in the output',
        'Helps assess the smoothness and speed of token generation.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, msecunits,
        'Most frequently occurring time between consecutive output tokens',
        'Helps assess the smoothness and speed of token generation.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Count of time taken between consecutive output tokens measures')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, msecunits,
        'Minimum time between generating consecutive tokens in the output')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, msecunits,
        'Maximum time between generating consecutive tokens in the output')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of time between consecutive output tokens')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within time between consecutive output tokens')

    pmns = 'guidellm.output_tokens_per_second.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, psecunits,
        'Average output tokens per second across all requests',
        'Provides insights into the servers performance and efficiency\n' +
        'in generating output tokens.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, psecunits,
        'Median prompt and output tokens per second across all requests',
        'Provides insights into the servers performance and efficiency\n' +
        'in generating output tokens.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, psecunits,
        'Mode of prompt and output tokens per second across all requests',
        'Provides insights into the servers performance and efficiency\n' +
        'in generating output tokens.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Count of output tokens per second measures')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, psecunits,
        'Minimum output tokens per second')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, psecunits,
        'Maximum output tokens per second')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of output tokens per second')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within output tokens per second')

    pmns = 'guidellm.tokens_per_second.total.'
    add_metric(pmns + 'mean', PM_TYPE_DOUBLE, psecunits,
        'Average prompt and output tokens per second across all requests',
        'Provides insights into the overall performance and efficiency\n' +
        'in processing both prompt and output tokens.')
    add_metric(pmns + 'median', PM_TYPE_DOUBLE, psecunits,
        'Median prompt and output tokens per second across all requests',
        'Provides insights into the overall performance and efficiency\n' +
        'in processing both prompt and output tokens.')
    add_metric(pmns + 'mode', PM_TYPE_DOUBLE, psecunits,
        'Mode of prompt and output tokens per second across all requests',
        'Provides insights into the overall performance and efficiency\n' +
        'in processing both prompt and output tokens.')
    add_metric(pmns + 'count', PM_TYPE_U64, unitone,
        'Count of prompt and output tokens per second measures')
    add_metric(pmns + 'min', PM_TYPE_DOUBLE, psecunits,
        'Minimum prompt and output tokens per second')
    add_metric(pmns + 'max', PM_TYPE_DOUBLE, psecunits,
        'Maximum prompt and output tokens per second')
    add_metric(pmns + 'std_dev', PM_TYPE_DOUBLE, nounit,
        'Standard deviation of prompt and output tokens per second')
    add_metric(pmns + 'variance', PM_TYPE_DOUBLE, nounit,
        'Variance within prompt and output tokens per second')

    return indom

def put_metric_values(pcp, values, instance) -> None:
    """ Scan through JSON data and extract values for each PCP metric;
        pmiPutValue inserts one value for one metric:inst at one time.
    """
    pmns = 'guidellm.'
    pcp.pmiPutValue(pmns + 'duration', instance, str(values['duration']))

    pmns = 'guidellm.run_stats.requests_made.'
    stats = values['run_stats']['requests_made']
    for metric in ['successful', 'errored', 'incomplete', 'total']:
        pcp.pmiPutValue(pmns + metric, instance, str(stats[metric]))

    pmns = 'guidellm.run_stats.'
    stats = values['run_stats']
    for metric in ['queued_time_avg', 'scheduled_time_delay_avg',
                   'worker_start_delay_avg', 'worker_time_avg',
                   'worker_start_time_targeted_delay_avg',
                   'request_start_time_delay_avg', 'request_time_delay_avg',
                   'request_start_time_targeted_delay_avg', 'request_time_avg']:
        pcp.pmiPutValue(pmns + metric, instance, str(stats[metric]))

    for group in ['requests_per_second', 'request_latency',
                  'time_to_first_token_ms', 'time_per_output_token_ms',
                  'inter_token_latency_ms', 'output_tokens_per_second',
                  'tokens_per_second']:
        pmns = 'guidellm.' + group + '.total.'
        stats = values['metrics'][group]['total']
        pcp.pmiPutValue(pmns + 'mean', instance, str(stats['mean']))
        pcp.pmiPutValue(pmns + 'mode', instance, str(stats['mode']))
        pcp.pmiPutValue(pmns + 'count', instance, str(stats['count']))
        pcp.pmiPutValue(pmns + 'median', instance, str(stats['median']))
        pcp.pmiPutValue(pmns + 'min', instance, str(stats['min']))
        pcp.pmiPutValue(pmns + 'max', instance, str(stats['max']))
        pcp.pmiPutValue(pmns + 'std_dev', instance, str(stats['std_dev']))
        pcp.pmiPutValue(pmns + 'variance', instance, str(stats['variance']))


def timestamp(timestring) -> datetime:
    return datetime.fromtimestamp(timestring, timezone.utc)


parser = argparse.ArgumentParser()
parser.add_argument("-v", "--verbose", help="Enable verbose progress diagnostics",
                    action="store_true")
parser.add_argument("-a", "--archive", help="Performance Co-Pilot archive output (timestamp if not specified)")
parser.add_argument("-H", "--hostname", help="Performance Co-Pilot archive hostname (run_id if not specified)")
parser.add_argument("results", help="JSON results from GuideLLM or Model Furnace")
args = parser.parse_args()

with open(args.results) as json_data:
    try:
        payload = json.load(json_data)
    except: # pylint: disable=bare-except
        print('Failed to load JSON document from:', args.results)
        sys.exit(1)
    instid = 0
    run_id = None

    # check if the results are wrapped with model furnace metadata
    try:
        result = payload['report'] # may generate a KeyError
    except KeyError:
        result = payload # default GuideLLM format

    for benchmark in result['benchmarks']:
        if not run_id:
            run_id = benchmark['run_id']
            if args.verbose:
                print('Run', run_id)
            # Create a new PCP archive
            if not args.hostname:
                args.hostname = run_id
            if not args.archive:
                args.archive = timestamp(benchmark['run_stats']['start_time'])
                args.archive = args.archive.strftime('%Y%m%d.%H.%M')
            log = pmi.pmiLogImport(args.archive)
            log.pmiSetHostname(args.hostname)
            log.pmiSetTimezone('UTC')
            log.pmiPutLabel(PM_LABEL_CONTEXT, 0, 0, 'guidellm_run_id', run_id)
            # extract some Model Furnace metadata as labels, when present
            for label in ['model', 'inference_server', 'accelerator_type']:
                value = payload.get(label)
                if value:
                    log.pmiPutLabel(PM_LABEL_CONTEXT, 0, 0, label, value)
            ids = register_metrics(log)

        instid = instid + 1
        instname = benchmark['id_']
        if args.verbose:
            print('[%d] %s' % (instid, instname))

        log.pmiAddInstance(ids, instname, instid)
        log.pmiPutLabel(PM_LABEL_CONTEXT, 0, 0, 'guidellm_id', instname)

        put_metric_values(log, benchmark, instname)
        log.pmiWrite(timestamp(benchmark['run_stats']['start_time']))

        put_metric_values(log, benchmark, instname)
        log.pmiWrite(timestamp(benchmark['run_stats']['end_time']))

        log.pmiPutMark() # end of benchmark iteration

    del log # flush the archive to persistent storage
